org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/typeHierarchy/TypeHierarchyView.java

```diff
@@ -11,32 +11,343 @@
 	public void createPartControl(Composite parent) {
-		final var filteredTree = new FilteredTree(parent, SWT.BORDER, new PatternFilter(), true, false) {
+		pagebook = new PageBook(parent, SWT.NULL);
+		splitter = new SashForm(pagebook, SWT.VERTICAL);
+		splitter.setLayoutData(new GridData(GridData.FILL_BOTH));
+		splitter.addControlListener(new ControlListener() {
+			@Override
+			public void controlMoved(ControlEvent e) {
+			}
+
+			@Override
+			public void controlResized(ControlEvent e) {
+				splitter.setOrientation(getBestOrientation());
+				splitter.layout();
+			}
+		});
+
+		treeViewer = getFilteredTree(splitter).getViewer();
+		treeViewer.setContentProvider(contentProvider);
+
+		treeViewer.setLabelProvider(new TypeHierarchyItemLabelProvider());
+
+		treeViewer.setUseHashlookup(true);
+		treeViewer.getControl().setEnabled(false);
+		treeViewer.addDoubleClickListener(new IDoubleClickListener() {
+			@Override
+			public void doubleClick(final DoubleClickEvent event) {
+				var selection = ((IStructuredSelection) event.getSelection()).getFirstElement();
+				if (selection instanceof TypeHierarchyItem item) {
+					try {
+						var symbolsContainer = cachedSymbols.get(new URI(item.getUri()));
+						if (symbolsContainer != null) {
+							LSPEclipseUtils.open(symbolsContainer.file.getLocationURI().toASCIIString(), item.getSelectionRange());
+						}
+					} catch (URISyntaxException e) {
+						LanguageServerPlugin.logError(e);
+					}
+				}
+			}
+		});
+		treeViewer.addSelectionChangedListener(new ISelectionChangedListener() {
+			@Override
+			public void selectionChanged(final SelectionChangedEvent event) {
+				onHierarchySelectionChanged(event);
+			}
+		});
+
+		memberViewForm = new ViewForm(splitter, SWT.NONE);
+		Control memberControl = createMemberControl(memberViewForm);
+		memberControl.setEnabled(false);
+		memberViewForm.setContent(memberControl);
+
+		memberLabel = new CLabel(memberViewForm, SWT.NONE);
+		memberViewForm.setTopLeft(memberLabel);
+
+		infoText = new Label(pagebook, SWT.TOP | SWT.LEFT | SWT.WRAP);
+		infoText.setText(Messages.TH_diplay_hint);
+		pagebook.showPage(infoText);
+
+		FileBuffers.getTextFileBufferManager().addFileBufferListener(fileBufferListener);
+	}
+
+	@Override
+	public void dispose() {
+		FileBuffers.getTextFileBufferManager().removeFileBufferListener(fileBufferListener);
+		cachedSymbols.forEach((uri, container) -> {container.dispose();});
+		super.dispose();
+	}
+
+	public void initialize(final IDocument document, final int offset) {
+		treeViewer.setInput(new HierarchyViewInput(document, offset));
+		pagebook.showPage(splitter);
+	}
+
+	private void onHierarchySelectionChanged(SelectionChangedEvent event) {
+		var selection = event.getSelection();
+		if (selection instanceof TreeSelection && !selection.isEmpty()) {
+			var element = ((TreeSelection) selection).getFirstElement();
+			if (element instanceof TypeHierarchyItem item) {
+				typeName = item.getName();
+				IFile file = null;
```

Comment:
could we move down the declaration of file to line 290? Just before the block where it is asigned?